### PR TITLE
Automatic use of C# primary constructors

### PR DESCRIPTION
**Pull request by roji**
Fully automated change. Note that almost all changes are in test classes only.

### CODE REVIEW

File: src/Shared/EnumerableExtensions.cs [hunk 1]
```diff
@@ -22,18 +22,11 @@ internal static class EnumerableExtensions
...
-        private readonly Func<T?, T?, bool> _func;
-
-        public DynamicEqualityComparer(Func<T?, T?, bool> func)
-        {
-            _func = func;
-        }
-
         public bool Equals(T? x, T? y)
-            => _func(x, y);
+            => func(x, y);
```
Comment (WeihanLi) [inline src/Shared/EnumerableExtensions.cs:29]: Want to mention that the `func` now is not `readonly` as before, though it may not matter much here

**Comment by WeihanLi** [comment]
Maybe we could add the config to editorconfig if we want to prefer primary-constructors style

```
# prefer primary constructor
csharp_style_prefer_primary_constructors = true:suggestion
```

File: src/Shared/EnumerableExtensions.cs (hunk 1 above)
Comment (roji) [reply src/Shared/EnumerableExtensions.cs:29]: Thanks, I agree it's not a big deal, but it is in other places. I've opened https://youtrack.jetbrains.com/issue/RIDER-104031/Convert-to-primary-constructor-incorrectly-turns-readonly-fields-to-non-readonly to flag this as a bug.

Comment (roji) [reply src/Shared/EnumerableExtensions.cs]: I've fixed this particular thing, and verified that in product code we don't have readonly fields converted to non-readonly; such instances remain in test code, where it doesn't seem too important. But if people have strong feelings about it, let me know.

**Comment by roji** [comment]
OK, used another method to apply the refactoring which does not suffer from the readonly problem mentioned above.

* Primary constructor refactoring done for tests only. Doing this in product code involves some manual work, deferring this for now.
* Refactoring to remove braces for empty types done for the entire solution:

```c#
// Instead of:
public class X
{
}

// We get:
public class X;
```

This is particularly nice in some test classes which become a single line.

**Review by AndriySvyryd — APPROVED** [review]
:sheep::it:

**Comment by lauxjpn** [comment]
This breaks or ability to override certain test cases.

Before PR, private class, public constructor:

```c#
private class Context9202 : DbContext
{
    public Context9202(DbContextOptions options)
        : base(options)
    {
    }
```

After PR, private class, _private_ constructor:

```c#
private class Context9202(DbContextOptions options) : DbContext(options)
{
```

**Comment by roji** [comment]
@lauxjpn if the class was private, wouldn't it already have been impossible to override it? In any case, even if Context9202 were protected you the test (presumably Include_collection_for_entity_with_owned_type_works?) would still use the base class rather than your overriding one I think...

If you need to override Include_collection_for_entity_with_owned_type_works, I think the way to do it is to duplicate Context9202 in your source code, and then override Include_collection_for_entity_with_owned_type_works, duplicating that too and using your own duplicated context..

Let me know if the above makes sense or I'm missing something.

**Comment by lauxjpn** [comment]
> if the class was private, wouldn't it already have been impossible to override it?

It's actually `Context18582` we are after, not `Context9202` (copied the wrong class name), but it shouldn't really matter since I think all those classes are effected.

We override `Projecting_correlated_collection_property_for_owned_entity` (that instantiates it), not the class itself.

> I think the way to do it is to duplicate Context9202 in your source code [...]

That workaround would work, but I think the intention should be to make overriding test cases as simple as possible and we would rather reuse the existing context class (that we don't need to change in any way and don't want to manually keep up-to-date with their original).

Making the former public constructor private seems to be an unintended side effect of the primary constructor refactoring of this PR and should be fixed for all affected classes.

A private/protected class with a public constructor should not be changed to use a primary constructor. Other providers are likely to hit the same issue in other classes (which is unnecessary since using a private primary constructor over a public regular one does not to seem to bring any gain here).

**Comment by roji** [comment]
I can indeed see that OwnedEntityTypeQueryBase.MyContext18582 was protected in 8.0, and is now private (and called Context18582); but this was done as part of #32542 by @maumar ([code](https://github.com/dotnet/efcore/pull/32542/files#diff-617374e301a3ca2d57f9cfc1675d06f6490ddddc647a5e536781ef4892a2e6e2R405)); in other words, I don't think it's really related to primary constructors, just about making sure that context types used by tests are protected rather than private, in case a provider needs to override/add some behavior there. Once the class itself is protected, unless I'm mistaken the primary constructor is also protected etc.

Do you want to submit a quick PR to make this context type (and any others which you need) protected? If not we can do it (but you're probably better-placed to know what you need).

**Comment by lauxjpn** [comment]
> I can indeed see that OwnedEntityTypeQueryBase.MyContext18582 was protected in 8.0, and is now private (and called Context18582); but this was done as part of #32542 by @maumar ([code](https://github.com/dotnet/efcore/pull/32542/files#diff-617374e301a3ca2d57f9cfc1675d06f6490ddddc647a5e536781ef4892a2e6e2R405)) [...]

Ah, you're right! It's #32542 that did the relevant change, not this PR here. Which means, that the automated changes worked as intended.

That makes things easier and narrows the scope significantly. I will follow up regarding `Context18582` etc. over at #32542 for context. Thanks!

**Pull request merged**